src/loader/mod.rs

```diff
@@ -0,0 +1,580 @@
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    BigEndianElfOnLittle,
+    CommandLineCopy,
+    CommandLineOverflow,
+    InvalidElfMagicNumber,
+    InvalidProgramHeaderSize,
+    InvalidProgramHeaderOffset,
+    InvalidProgramHeaderAddress,
+    InvalidEntryAddress,
+    InvalidBzImage,
+    InvalidKernelStartAddress,
+    InitrdImageSizeTooLarge,
+    ReadElfHeader,
+    ReadKernelImage,
+    ReadProgramHeader,
+    ReadBzImageHeader,
+    ReadBzImageCompressedKernel,
+    ReadInitrdImage,
+    SeekKernelStart,
+    SeekElfStart,
+    SeekProgramHeader,
+    SeekBzImageEnd,
+    SeekBzImageHeader,
+    SeekBzImageCompressedKernel,
+    SeekInitrdImage,
+}
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl error::Error for Error {
+    fn description(&self) -> &str {
+        match self {
+            Error::BigEndianElfOnLittle => {
+                "Trying to load big-endian binary on little-endian machine"
+            }
+            Error::CommandLineCopy => "Failed writing command line to guest memory",
+            Error::CommandLineOverflow => "Command line overflowed guest memory",
+            Error::InvalidElfMagicNumber => "Invalid Elf magic number",
+            Error::InvalidProgramHeaderSize => "Invalid program header size",
+            Error::InvalidProgramHeaderOffset => "Invalid program header offset",
+            Error::InvalidProgramHeaderAddress => "Invalid Program Header Address",
+            Error::InvalidEntryAddress => "Invalid entry address",
+            Error::InvalidBzImage => "Invalid bzImage",
+            Error::InvalidKernelStartAddress => "Invalid kernel start address",
+            Error::InitrdImageSizeTooLarge => "Initrd image size too large",
+            Error::ReadElfHeader => "Unable to read elf header",
+            Error::ReadKernelImage => "Unable to read kernel image",
+            Error::ReadProgramHeader => "Unable to read program header",
+            Error::ReadBzImageHeader => "Unable to read bzImage header",
+            Error::ReadBzImageCompressedKernel => "Unable to read bzImage compressed kernel",
+            Error::ReadInitrdImage => "Unable to read initrd image",
+            Error::SeekKernelStart => "Unable to seek to kernel start",
+            Error::SeekElfStart => "Unable to seek to elf start",
+            Error::SeekProgramHeader => "Unable to seek to program header",
+            Error::SeekBzImageEnd => "Unable to seek bzImage end",
+            Error::SeekBzImageHeader => "Unable to seek bzImage header",
+            Error::SeekBzImageCompressedKernel => "Unable to seek bzImage compressed kernel",
+            Error::SeekInitrdImage => "Unable to seek initrd image",
+        }
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Kernel Loader Error: {}", Error::description(self))
+    }
+}
+
+/// * `kernel_load` - The actual `guest_mem` address where kernel image is loaded start.
```

Comment:
The explanation of each field in the structure must be written before each field instead of at the beginning of the struct definition.

Can you also please rephrase the details to include the fact that kernel_load and kernel_end are addresses/offsets? Also, I wouldn't use `guest_mem` in the explanation because as an outside user of this crate you don't know what the `guest_mem` variable corresponds to.

```rust
#[derive(Debug, Default, Copy, Clone, PartialEq)]
pub struct KernelLoaderResult {
    /// Address in the guest memory where the kernel image is loaded.
    pub kernel_load: GuestAddress,
   /// Offset in guest memory corresponding to the end of the kernel image. TODO: add details about initrd. 
    pub kernel_end: GuestUsize,
    /// Optional header that specifies "TODO something something".
   /// This field is only used when loading bzimage and it is necessary because "something something"
    pub setup_header: Option<bootparam::setup_header>,
}
```

I don't know how bzimage works just yet, so i just wrote something something :))

Reply:
Nice!